### ISSUE TITLE
Unclear if ELMo module usable with text_field_embedders?

### ISSUE THREAD

**Issue by ethch18** [allenai/allennlp#1262]
The [ELMo Readme](https://github.com/allenai/allennlp/blob/master/tutorials/how_to/elmo.md) contains an example for simple cases in which token embeddings and ELMo embeddings (via `ElmoTokenEmbedder`) are used side-by-side.  I'm trying to do the same, except with the ELMo module, for a textual entailment model that I'm writing.  Given that `TextFieldEmbedder`s need a corresponding entry for every `TokenIndexer` used by the `DatasetReader`, and the ELMo module is not run as part of the `TextFieldEmbedder`, is this still possible?  Thanks!

For more context, here's what I have in terms of config:
```
"dataset_reader": {
  "type": "snli",
  "token_indexers": {
    "elmo": {
      "type": "elmo_characters"
    },
    "tokens": {
      "type": "single_id",
      "lowercase_tokens" : true
    }
  },
  "tokenizer": {
    "end_tokens": ["@@NULL@@"]
  }
},
[...]
"model": {
  [...]
  "text_field_embedder": {
    "tokens": {
      "type": "embedding",
      "projection_dim": 200,
      "pretrained_file": "https://s3-us-west-2.amazonaws.com/allennlp/datasets/glove/glove.6B.300d.txt.gz",
      "embedding_dim": 300,
      "trainable": false
    },
    "elmo": {}
  },
  [...]
}
```

Since the second `"elmo"` entry is required by the `TextFieldEmbedder`, my goal is to extract (for example) only the `tokens` part of `self._text_field_embedder(premise)`, and then concatenate them with (for example) `self._elmo_module(premise["elmo"])["elmo_representations"]`.

**Comment by matt-gardner** [comment]
If you pass the right flags in the "text_field_embedder" key, then the ELMo module _will_ be run as part of the `TextFieldEmbedder`.  Where you have `"elmo": {}`, you should instead have something like this:

```
"elmo": {
    "type": "elmo_token_embedder",
    "options_file": "https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_options.json",
    "weight_file": "https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_weights.hdf5",
    "do_layer_norm": false,
    "dropout": 0.5
  }
```

You shouldn't need any other changes to your code or config file.

**Comment by ethch18** [comment]
Thanks!  I'm actually trying to use multiple layers of ELMo, which is why (if I'm understanding correctly) I can't use the `ElmoTokenEmbedder`, so I wanted to see if there was a way to use the actual `allennlp.modules.elmo.Elmo` module in its place.

**Comment by nelson-liu** [comment]
maybe i misinterpreted, but i think @ethch18 wants to get the elmo embeddings programmatically _without_ using the `ElmoEmbedder` and using the `Elmo` class instead (e.g., for using in other parts of the model)?

if so:

There's a moderately hacky solution at https://github.com/allenai/allennlp/pull/1253/files#diff-32c000dbed2d9b2c1f1289f2971eb928R205 (the whole PR in general and its comments might be useful to you)

In short, the `elmo` key in the tokens is popped off before being passed to the embedder, and then the elmo representations are calculated separately and added if applicable. The model throws an error if elmo is specified in the embedder, and there are two flags for where to use elmo within the model

If all you want to do with ELMo is append it to the input, you should do what matt said above.

**Comment by ethch18** [comment]
Thanks @nelson-liu, that helps a lot!
